### PR TITLE
Taxonomies: Retain flat term order

### PR DESCRIPTION
**Pull request by ockham**
## What?
See https://github.com/WordPress/gutenberg/issues/65052.

## Why?
To retain the order in which "flat terms" (such as tags) are stored for a given post.

## How?
Sort the terms obtained via `getEntityRecords` by the array of term IDs that reflects the order in which they were added by the user to the current post.

## TODO

- [ ] We might want to respect the `sort` setting that's provided to [`register_taxonomy`](https://developer.wordpress.org/reference/functions/register_taxonomy/) to determine whether terms should be stored in the order they're given by the user (if `sort === true`, somewhat counter-intuitively), or if we should continue to sort them alphabetically.

> Whether terms in this taxonomy should be sorted in the order they are provided to `wp_set_object_terms()`. Default `null` which equates to `false`.

## Testing Instructions
Add the following code (e.g. to your theme's `functions.php`) and use the "Actors" panel in the block inspector to assign a number of actor names to a given post. Note that the order is kept (unlike previously, when they were re-ordered in alphabetical order). _Note however that when **saving** the post, the terms are still sorted alphabetically, and their previous order is lost. See https://github.com/WordPress/wordpress-develop/pull/7287 for a WIP fix._

```php
function register_taxonomies() {
	/**
	 * Taxonomy: Actors.
	 */

	 $labels = array(
		'name'          => __( 'Actors' ),
		'singular_name' => __( 'Actor' ),
		'add_new_item'  => __( 'Add Actor' ),
		'new_item_name' => __( 'New Actor' ),
	);

	$args = array(
		'label'                 => __( 'Cast' ),
		'labels'                => $labels,
		'public'                => true,
		'publicly_queryable'    => true,
		'hierarchical'          => false,
		'show_ui'               => true,
		'show_in_menu'          => true,
		'show_in_nav_menus'     => true,
		'query_var'             => true,
		'show_admin_column'     => false,
		'show_in_rest'          => true,
		'show_tagcloud'         => false,
		'rest_base'             => 'actors',
		'rest_controller_class' => 'WP_REST_Terms_Controller',
		'rest_namespace'        => 'wp/v2',
		'show_in_quick_edit'    => false,
		'show_in_graphql'       => false,
		'sort'                  => true, // Note this.
	);

	register_taxonomy( 'actors', array( 'post' ), $args );
}

add_action( 'init', 'register_taxonomies', 0 );
```

## Screenshots or screencast 

| Before | After |
|-------|-------|
| flat-terms-before| flat-terms-after |

### CODE REVIEW

File: packages/editor/src/components/post-taxonomies/flat-term-selector.js [hunk 1]
```diff
@@ -116,8 +120,10 @@ export function FlatTermSelector( { slug, __nextHasNoMarginBottom } ) {
 					: false,
 				taxonomy: _taxonomy,
 				termIds: _termIds,
-				terms: _termIds?.length
-					? getEntityRecords( 'taxonomy', slug, query )
+				terms: _terms
+					? _termIds.map( ( termId ) =>
+							_terms.find( ( term ) => term.id === termId )
+					  )
```
Comment (ockham) [inline packages/editor/src/components/post-taxonomies/flat-term-selector.js:126]: We might want to use a more efficient way of sorting here, i.e. via [`Array.prototype.sort`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort), and a suitable comparison function.

OTOH, using `map()` is pretty intuitive here. Also, `sort()` sorts in-place, i.e. retains referential equality, which isn't really desirable here.

**Comment by ockham** [comment]
The server-side part that would be needed for this seems to be a bit more complex. I'll file an issue to accompany this PR and will track my findings there.

**Comment by jsnajdr** [comment]
There's one additional caveat when it comes to sorting terms in a "natural" order. WooCommerce is adding several taxonomies, e.g., Product Categories, and lets users to sort the terms using drag&drop handles:

Screenshot 2024-10-03 at 12 50 00

This order is sorted as an `order` field in term meta and is supposed to be the natural order in which the terms are displayed on frontend, when there is no other `order_by` specified.

Therefore, the "natural" order is not always by ID, how the terms were chronologically added.

Besides Woo, there are also [specialized popular plugins](https://wordpress.org/plugins/taxonomy-terms-order/) that add the same feature to any taxonomy.

This Woo feature currently doesn't work with Gutenberg. It probably works when the terms are listed in a classic theme using a PHP function, but i) the REST API doesn't respect this order ii) neither the Gutenberg categories block respects that when rendered by the server for the frontend. Note in the following screenshots that the "Tables" category should be the first, but it isn't:

Gutenberg:
Screenshot 2024-10-03 at 12 54 45

Frontend:
Screenshot 2024-10-03 at 12 55 12

I'm sharing this here because it seems to me that all these problems are closely related and a 100% correct patch will solve them all at once.

**Comment by ockham** [comment]
I finally managed to get back to this. I've given this some more thought, and I think one of the key problems here is that setting `sort` to `true` in [`register_taxonomy`](https://developer.wordpress.org/reference/functions/register_taxonomy/) only affects the order in which terms are _stored_; it doesn't make any statements about the order in which they're sorted when _read_ from the DB.

This is complicated by the fact that high-level functions such as [`get_the_terms`](https://developer.wordpress.org/reference/functions/get_the_terms/), [`get_the_term_list`](https://developer.wordpress.org/reference/functions/get_the_term_list/) do not support any arguments to modify the order in which terms are listed (or any arguments to impact how they're presented, really); you have to use the lower-level [`wp_get_object_terms`](https://developer.wordpress.org/reference/functions/wp_get_object_terms/) to specify those. This seems to be (at least partly) because of caching; by locking down `get_the_terms` and friends to only one canonical form, caching becomes much easier than if it'd had to deal with different possible ways of ordering terms.

It seems that the recommended way of defining a different sort order is at `register_taxonomy` time, which supports an argument called `args`:

> Array of arguments to automatically use inside `wp_get_object_terms()` for this taxonomy.

For example, `args => array( 'orderby' => 'term_order' )` will return terms in the order they were saved (if `sort` is set to `true`). This is found in [these two unit tests](https://github.com/WordPress/wordpress-develop/blob/11e4c7c5dd8c0366e0f911995881a6ca9a2323e5/tests/phpunit/tests/term/query.php#L519-L577) in Core that actually set `sort` to `true`.

(As a corollary, I believe that using `args => array( 'orderby' => /* Woo order term meta */ )` in Woo's `register_taxonomy` call should fix the issue that @jsnajdr described. Can you give that a try and confirm, Jarda? 😊)

---

This raises the question if `'sort' => true` should implicitly set `args => array( 'orderby' => 'term_order' )`. Unit tests aside, Core doesn't really provide an answer for this, as `'sort' => true` isn't used for any of its own taxonomies -- after all, categories and tags are rendered in alphabetic order.

I'm torn on this. I've started trying it out in https://github.com/WordPress/wordpress-develop/pull/7287. It's working for the REST API (thus retaining term order in the editor sidebar panel), but not on the frontend, which goes through a different pipeline (via `get_the_terms()`. It's been  a while since I've touched that code, but I think the biggest hurdle there was to default to `args => array( 'orderby' => 'term_order' )` if `sort` is set to `true`, while still allowing the user to override that behavior (e.g. by setting `args[ 'orderby' ]` to a different value); the latter was hard if not impossible to do IIRC.

I'd be very curious to hear your thoughts on this @jsnajdr!

**Comment by ockham** [comment]
One more interesting find: The `WP_Taxonomy` class has a `sort` field (reflecting the `register_taxonomy` argument of the same name), [whose PHPDoc says](https://github.com/WordPress/wordpress-develop/blob/11e4c7c5dd8c0366e0f911995881a6ca9a2323e5/src/wp-includes/class-wp-taxonomy.php#L251):

```
Use this in combination with `'orderby' => 'term_order'` when fetching terms.
```

**Comment by ockham** [comment]
I think I have an idea how to proceed.

- I'll probably close https://github.com/WordPress/wordpress-develop/pull/7287 -- at least for now -- since the user can achieve the same effect by setting both `sort` to `true`, and `args` to `array( 'orderby' => 'term_order' )` (as explained in [this comment](https://github.com/WordPress/gutenberg/pull/64471#issuecomment-2485326942)).
- I'll expose `args` -- or at least the `orderby` arg -- in the taxonomies controller, and will then update the query in `packages/editor/src/components/post-taxonomies/flat-term-selector.js` to set `orderby` to whatever the taxonomies controller returns for that. (I'll do this as part of this PR, i.e. with an override class for the Taxonomies controller in the 6.8 compat layer).

**Comment by ockham** [comment]
Something doesn't sit quite right with me about [querying the sort order from the network only to include in another network request](https://github.com/WordPress/gutenberg/pull/64471#issuecomment-2486086381) -- it sounds like the server should have all the information it needs, so I'll try to think about this some more.

The big question is: If a taxonomy is registered with both `sort` set to `true`, and `args => array( 'orderby' => 'term_order' )` -- why are terms still being ordered alphabetically when a new term is entered into the terms input panel? Shouldn't the endpoint respect the `args` (and thus the `orderby`) that were specified when the taxonomy was registered? 🤔

**Comment by ockham** [comment]
> The big question is: If a taxonomy is registered with both `sort` set to `true`, and `args => array( 'orderby' => 'term_order' )` -- why are terms still being ordered alphabetically when a new term is entered into the terms input panel? Shouldn't the endpoint respect the `args` (and thus the `orderby`) that were specified when the taxonomy was registered? 🤔

Getting closer to an answer. Turns out that to get the list of terms, [the controller uses](https://github.com/WordPress/wordpress-develop/blob/193f6eb26e65216576a10492dd4103044c3a079c/src/wp-includes/rest-api/endpoints/class-wp-rest-terms-controller.php#L320-L327) either `wp_get_object_terms` (if a `post` ID is specified by the network request), or `get_terms` (if there isn't). And while `wp_get_object_terms` [takes the `args` from `register_taxonomy` into account](https://github.com/WordPress/wordpress-develop/blob/193f6eb26e65216576a10492dd4103044c3a079c/src/wp-includes/taxonomy.php#L2292-L2310), it appears that [`get_terms`](https://github.com/WordPress/wordpress-develop/blob/193f6eb26e65216576a10492dd4103044c3a079c/src/wp-includes/taxonomy.php#L1316) doesn't (although I haven't yet checked the entire callstack for the latter).

And the queries that `flat-term-selector.js` actually don't contain a post ID! So if my theory is correct, there might be a potential quick fix by including the latter in the query.

(We might still want to consider why `get_terms` ignores the taxonomy's default `args`, and if we should change that.)

**Comment by ockham** [comment]
> So if my theory is correct, there might be a potential quick fix by including the latter in the query.

Ah, we can't really do that. The problem is that when a new term is entered by the user, it is created (via a POST request), but not yet associated with the current post. This is by design: We only want that to happen once the user clicks the 'Save' button in order to persist their changes. (Otherwise, the new term would show up on that post, even if the user ends up never saving the post.)

So it looks like we need to explore the other avenue:

> We might still want to consider why `get_terms` ignores the taxonomy's default args, and if we should change that.

**Comment by ockham** [comment]
> > We might still want to consider why `get_terms` ignores the taxonomy's default args, and if we should change that.

The following works (for a taxonomy called `actors`) 🎉 

```diff
diff --git a/lib/compat/wordpress-6.8/rest-api.php b/lib/compat/wordpress-6.8/rest-api.php
index da1b657cda0..3c465e2a487 100644
--- a/lib/compat/wordpress-6.8/rest-api.php
+++ b/lib/compat/wordpress-6.8/rest-api.php
@@ -20,3 +20,13 @@ if ( ! function_exists( 'gutenberg_add_post_type_rendering_mode' ) ) {
        }
 }
 add_action( 'rest_api_init', 'gutenberg_add_post_type_rendering_mode' );
+
+function gutenberg_respect_taxonomy_default_args( $args ) {
+       $taxonomy = $args['taxonomy'];
+       $t = get_taxonomy( $taxonomy );
+       if ( isset( $t->args ) && is_array( $t->args ) ) {
+               $args = array_merge( $args, $t->args );
+       }
+       return $args;
+}
+add_filter( 'rest_actors_query', 'gutenberg_respect_taxonomy_default_args' );
```

Now I'll have to see how I can generalize it 🙂

**Comment by ockham** [comment]
Closing in favor of https://github.com/WordPress/gutenberg/pull/67154.

**Comment by jsnajdr** [comment]
> As a corollary, I believe that using `args => array( 'orderby' => /* Woo order term meta */ )` in Woo's `register_taxonomy` call should fix the issue that @jsnajdr described. Can you give that a try and confirm, Jarda?

Yes, Woo stores the custom order in an `order` field in term meta and WordPress has a native support for sorting by a meta value:
```
args => array(
  'orderby' => 'meta_value_num',
  'meta_key' => 'order'
)
```
Woo doesn't add these default args in `register_taxonomy`, but instead adds them in various filters and custom wrappers. So, listing the terms for the `wp-admin/edit-tags.php` page works because the query goes through a filter. I'm not so sure about the more recently added REST queries used by Gutenberg. If they don't work, it's a Woo bug where Woo should add or extend a filter that sets the `orderby` args.

**Pull request closed without merging**